partiql-planner/src/main/kotlin/org/partiql/planner/internal/typer/ExcludeUtils.kt

```diff
@@ -0,0 +1,159 @@
+import org.partiql.errors.ProblemCallback
+import org.partiql.errors.UNKNOWN_PROBLEM_LOCATION
+import org.partiql.planner.PlanningProblemDetails
+import org.partiql.planner.internal.ir.Identifier
+import org.partiql.planner.internal.ir.Identifier.CaseSensitivity
+import org.partiql.planner.internal.ir.Rel
+import org.partiql.planner.internal.ir.Rex
+import org.partiql.types.AnyOfType
+import org.partiql.types.CollectionType
+import org.partiql.types.StaticType
+import org.partiql.types.StructType
+
+/**
+ * Checks for every exclude path in [paths], that a value in the [bindings] is excluded. If there is no value excluded,
+ * a warning is added to the [onProblem] callback.
+ */
+internal fun checkForInvalidExcludePaths(bindings: List<Rel.Binding>, paths: List<Rel.Op.Exclude.Item>, onProblem: ProblemCallback) {
+    paths.forEach { excludePath ->
+        val root = excludePath.root
+        val steps = excludePath.steps
+        val excludesSomething = bindings.any { binding ->
+            when (root) {
+                is Rex.Op.Var.Unresolved -> {
+                    when (val id = root.identifier) {
+                        is Identifier.Symbol -> {
+                            if (id.isEquivalentTo(binding.name)) {
+                                binding.type.checkExclude(steps)
+                            } else {
+                                false
+                            }
+                        }
+                        is Identifier.Qualified -> {
+                            if (id.root.isEquivalentTo(binding.name)) {
+                                binding.type.checkExclude(steps)
+                            } else {
+                                false
+                            }
+                        }
+                    }
+                }
+                is Rex.Op.Var.Resolved -> false
+            }
+        }
+        // If nothing is excluded by `excludePath`, add a warning
+        if (!excludesSomething) {
+            onProblem(
+                Problem(
+                    sourceLocation = UNKNOWN_PROBLEM_LOCATION,
+                    details = PlanningProblemDetails.InvalidExcludePath(
+                        excludePath.toProblemString()
+                    )
+                )
+            )
+        }
+    }
+}
+
+/**
+ * Checks whether [steps] will exclude a value from [this] [StaticType].
+ */
+internal fun StaticType.checkExclude(steps: List<Rel.Op.Exclude.Step>): Boolean {
+    return when (this) {
+        is StructType -> this.checkExclude(steps)
+        is CollectionType -> this.checkExclude(steps)
+        is AnyOfType -> this.types.any { it.checkExclude(steps) }
+        else -> false
+    }
+}
+
+/**
+ * Checks whether [steps] will exclude a value from [this] [StructType].
+ */
+internal fun StructType.checkExclude(steps: List<Rel.Op.Exclude.Step>): Boolean {
+    // Ignore open structs
+    if (!this.contentClosed) {
+        return true
+    }
+    val step = steps.first()
```

Comment:
Check step size before calling first. We don't want this to blow up if the steps list is empty (just in case) even though it _should_ never be empty.